docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+- 2020-11-09: Initial version (@erikgrinaker)
+
+## Context
+
+In [ADR 061](adr-061-p2p-refactor-scope.md) we decided to refactor the peer-to-peer (P2P) networking stack. The first phase is to redesign and refactor the internal P2P architecture, while retaining protocol compatibility as far as possible.
+
+## Alternative Approaches
+
+Several variations of the proposed design were considered, including e.g. calling interface methods instead of passing messages (like the current architecture), merging channels with streams, exposing the internal peer data structure to reactors, being message format-agnostic via arbitrary codecs, and so on. This design was chosen because it has very loose coupling, is simpler to reason about and more convenient to use, avoids race conditions and lock contention for internal data structures, gives reactors better control of message ordering and processing semantics, and allows for QoS scheduling and backpressure in a very natural way.
+
+There were also proposals to use LibP2P instead of maintaining our own P2P stack, which was rejected (for now) in [ADR 061](adr-061-p2p-refactor-scope.md).
+
+## Decision
+
+The P2P stack will be redesigned as a message-oriented architecture, primarily relying on Go channels for communication and scheduling. It will use IO stream transports to exchange raw bytes with individual peers, bidirectional peer-addressable channels to send and receive Protobuf messages, and a router to route messages between reactors and peers. Message passing is asynchronous with at-most-once delivery.
+
+## Detailed Design
+
+This ADR is primarily concerned with the architecture and interfaces of the P2P stack, not implementation details. Separate ADRs may be submitted for individual components, since implementation may be non-trivial. The interfaces described here should therefore be considered a rough architecture outline, not a complete and final design.
+
+Primary design objectives have been:
+
+* Loose coupling between components, for a simpler, more robust, and test-friendly architecture.
+* Pluggable transports (not necessarily networked).
+* Better scheduling of messages, with improved prioritization, backpressure, and performance.
+* Centralized peer lifecycle and connection management.
+* Better peer address detection, advertisement, and exchange.
+* Backwards compatibility at the wire-level with current P2P network protocols.
+
+The main abstractions in the new stack are:
+
+* `peer`: A node in the network, uniquely identified by a `PeerID` and stored in a `peerStore`.
+* `Transport`: An arbitrary mechanism to exchange bytes with a peer using IO `Stream`s across a `Connection`.
+* `Channel`: A bidirectional channel to asynchronously exchange Protobuf messages with peers addressed with `PeerID`.
+* `Router`: Maintains transport connections to relevant peers and routes channel messages.
+* Reactor: A design pattern loosely defined as "something which listens on a channel and reacts to messages".
```

Comment:
I think that Handler could've been a better name for this thing

Reply:
Yeah, handler works. I felt like reactor is more in line with the current stack, and would make it easier for people to map the concepts (which is why I've kept most of the other names too). Would be interested to head what others think.

Reply:
I'm generally not in favor of the standalone name reactor, because the semantic mapping between "thing that reacts to events" is something I never think about. (The type of event isn't that clear from the name, and I always think of reactor as in 'nuclear reactor' when hearing the name)

Handler's also not a significantly better name imo ('thing that handles' is similarly vague, but at least more common). Perhaps calling it a p2p Reactor and only referring to it as such may help?

Reply:
I'm terrible with names, so don't have any better suggestions really, but I feel like handler (as a shorthand for message handler) is slightly better than reactor in that it's at least obvious what it does: handler => handles stuff, reactor => nuclear reactor wtf?

Reply:
lol lol lol I also thought nuclear reactor when I first joined

Reply:
I've discussed this with a few people, and since no-one has been able to come up with significantly better names we're sticking with "reactor" and "channel" for now as these terms are already fairly ingrained. We can absolutely change them if we come up with anything better.